### PR TITLE
Sms simulator (front-end)

### LINKED ISSUE #1697: Session simulator frontend
Based on https://github.com/instedd/surveda/pull/1696 implement the UI on https://projects.invisionapp.com/d/main#/console/8374235/385278474/preview .

User should be able to do a full page reload to start a new session.

The chat component WIP is in `feature/sms-simulator` branch
The submission and state controls already exist but they might need to be decoupled from the store since in the simulator the response from the server should be used.

There is no need to store the simulated session in the store. It can all be kept in the state of the page component.

### PR DESCRIPTION
**Pull request by aeatencio**
This is how it looks:

active
random section skipped
random section first
refused

Fix #1697

### CODE REVIEW

File: web/static/js/components/questionnaires/QuestionnaireTitle.jsx [hunk 1]
```diff
@@ -34,14 +34,20 @@ class QuestionnaireTitle extends Component {
...
-        <EditableDescriptionLabel description={questionnaire.description} emptyText={t('Add description')} onSubmit={(value) => { this.handleSubmitDescription(value) }} readOnly={readOnly} />
+        <EditableDescriptionLabel description={questionnaire.description} emptyText={readOnly ? '' : t('Add description')} onSubmit={(value) => { this.handleSubmitDescription(value) }} readOnly={readOnly} />
       </div>
     )
   }
 }
 
 const mapStateToProps = (state, ownProps) => ({
-  readOnly: state.project && state.project.data ? state.project.data.readOnly : true
+  readOnly: typeof ownProps.readOnly === 'boolean'
```
Comment (bcardiff) [inline web/static/js/components/questionnaires/QuestionnaireTitle.jsx:44]: why this change?

File: web/static/js/components/questionnaires/QuestionnaireTitle.jsx [hunk 2]
```diff
@@ -34,14 +34,20 @@ class QuestionnaireTitle extends Component {
     return (
       <div className='title-container'>
         <EditableTitleLabel title={questionnaire.name} onSubmit={(value) => { this.handleSubmit(value) }} emptyText={t('Untitled questionnaire')} readOnly={readOnly} more={<QuestionnaireMenu />} />
-        <EditableDescriptionLabel description={questionnaire.description} emptyText={t('Add description')} onSubmit={(value) => { this.handleSubmitDescription(value) }} readOnly={readOnly} />
+        <EditableDescriptionLabel description={questionnaire.description} emptyText={readOnly ? '' : t('Add description')} onSubmit={(value) => { this.handleSubmitDescription(value) }} readOnly={readOnly} />
```
Comment (bcardiff) [inline web/static/js/components/questionnaires/QuestionnaireTitle.jsx:37]: If it is readonly it should not even have an `onSubmit` attribute, right?

File: web/static/js/components/simulation/SimulationSteps.jsx [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  }
+
+  wasSkipped = (stepId: string): boolean => {
+    if (this.isActive(stepId) || this.isCompleted(stepId)) return false
+    const { currentStepId } = this.props
+    if (currentStepId) {
+      const { headedBySectionSteps } = this.state
+      return headedBySectionSteps.findIndex(step => step.id == stepId) < headedBySectionSteps.findIndex(step => step.id == currentStepId)
+    } else {
+      return true
```
Comment (bcardiff) [inline web/static/js/components/simulation/SimulationSteps.jsx:92]: Why if there is no current step all steps are marked as "skipped"?

The section headers are also marked as completed or they render different that question steps.

File: web/static/js/components/simulation/QuestionnaireSimulation.jsx [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+  state = {
+    simulation: null
+  }
+
+  fetchQuestionnaireForTitle = () => {
+    const { projectId, questionnaireId } = this.props
+    // TODO: this component doesn't do this fetch for itself
+    // It does it just because the title component above needs the questionnaire being in the redux store to render properly
+    // But for some reason that component isn't able to get the projectId and questionnaireId params properly from the url
+    this.props.questionnaireActions.fetchQuestionnaireIfNeeded(projectId, questionnaireId)
```
Comment (bcardiff) [inline web/static/js/components/simulation/QuestionnaireSimulation.jsx:53]: Aren't you missing a `withRouter` in the title component?

File: web/static/js/components/questionnaires/QuestionnaireTitle.jsx (hunk 1 above)
Comment (aeatencio) [reply web/static/js/components/questionnaires/QuestionnaireTitle.jsx:44]: Because we don't want the title to be changed from the simulation page.

It's related to [this other change](https://github.com/instedd/surveda/pull/1698/commits/621d2946311c1e0361cf60df7cd00916bd456ddb#diff-628a34867fbc251edf9230cc7c96e629R66).

So, the intention here was: if the component's user explicitly decides to set it read-only (or the opposite) prioritize that.

File: web/static/js/components/simulation/SimulationSteps.jsx (hunk 3 above)
Comment (aeatencio) [reply web/static/js/components/simulation/SimulationSteps.jsx:92]: The truth is that the case I'm covering defaulting to `true` here is this (simulation ended)

simulation ended

I'm not really sure if defaulting it to `true` is the best solution. What do you think?

Regarding the section headers, they are always being classified as skipped (because of the above default). I think visually they can be OK, but I'm pretty sure the semantic isn't good. I'll try to make it better and let you know if I find that I can't.

File: web/static/js/components/simulation/QuestionnaireSimulation.jsx (hunk 4 above)
Comment (aeatencio) [reply web/static/js/components/simulation/QuestionnaireSimulation.jsx:53]: I'm almost sure it wasn't because of that, but I'll try it again and let you know! Thanks!

File: web/static/js/components/questionnaires/QuestionnaireTitle.jsx (hunk 1 above)
Comment (bcardiff) [reply web/static/js/components/questionnaires/QuestionnaireTitle.jsx]: ok

File: web/static/js/components/simulation/SimulationSteps.jsx (hunk 3 above)
Comment (bcardiff) [reply web/static/js/components/simulation/SimulationSteps.jsx:92]: By default questions should not be marked as skipped.

Even if visually they look the same. For now it's fine if by default pending and skipped questions look the same, but let's not add a css class of skipped to default so we can iterate later on the styling if needed.

I don't follow why the headers of sections are marked as skipped. I would expect for them to be in the `headedBySectionSteps` list.

Maybe the step representing the header can also be marked as active/completed/pending=default based on the following rules:

* if the current step is of the section, then the section is active (note, the question step itself will also be active)
* if the current step is NOT of the section, and there are already answers within the section, then section is completed.
* otherwise, is pending=default.

Between these changes and knowing the step type, we can tweak later the styling to show progress in sections nicely I think. Again, I don't mind the styling in this PR.

File: web/static/js/components/simulation/QuestionnaireSimulation.jsx (hunk 4 above)
Comment (aeatencio) [reply web/static/js/components/simulation/QuestionnaireSimulation.jsx]: OK, yes, adding `withRouter` I was able to access the params.

Now, it's solved [here](https://github.com/instedd/surveda/pull/1698/commits/40b62e16f91fc789f5458a87bcfc0bb2b2559922). Thanks!!!

File: web/static/js/components/simulation/QuestionnaireSimulation.jsx [hunk 5]
```diff
@@ -33,18 +34,27 @@ type Props = {
...
+  questionnaireActions: Object,
+  t: Function
 }
 
 type State = {
   simulation: ?Simulation
 }
 
-class QuestionnaireSimulation extends Component<Props, State> {
+const QuestionnaireSimulation = translate()(class extends Component<Props, State> {
```
Comment (bcardiff) [inline web/static/js/components/simulation/QuestionnaireSimulation.jsx:42]: connect with translate at the bottom of the file with the export default.

File: web/static/js/components/questionnaires/QuestionnaireTitle.jsx (hunk 2 above)
Comment (aeatencio) [reply web/static/js/components/questionnaires/QuestionnaireTitle.jsx]: I tried, but `onSubmit` is required. Should we change that?

Comment (bcardiff) [reply web/static/js/components/questionnaires/QuestionnaireTitle.jsx]: ok, no worries.

File: web/static/js/components/simulation/SimulationSteps.jsx (hunk 3 above)
Comment (aeatencio) [reply web/static/js/components/simulation/SimulationSteps.jsx]: I think this problem was solved in [this commit](https://github.com/instedd/surveda/pull/1698/commits/f9981048f94c70b64358c328fee36bb7d2d9c489)

File: web/static/js/components/simulation/QuestionnaireSimulation.jsx (hunk 5 above)
Comment (aeatencio) [reply web/static/js/components/simulation/QuestionnaireSimulation.jsx]: Done!

**Comment by aeatencio** [comment]
@bcardiff, now I think it's ready for review! 🤞

File: web/static/js/components/questionnaires/QuestionnaireTitle.jsx [hunk 6]
```diff
@@ -30,18 +40,35 @@ class QuestionnaireTitle extends Component {
...
     )
   }
 }
 
-const mapStateToProps = (state, ownProps) => ({
-  readOnly: state.project && state.project.data ? state.project.data.readOnly : true
-})
-
-export default translate()(connect(mapStateToProps)(withQuestionnaire(QuestionnaireTitle)))
+export default translate()(withRouter(connect(mapStateToProps)(QuestionnaireTitle)))
```
Comment (bcardiff) [inline web/static/js/components/questionnaires/QuestionnaireTitle.jsx:74]: I'm checking that `withQuestionnaire` is responsible of putting the quex in redux store into de property of the component. Why this need to change? Other usages of `fetchQuestionnaireIfNeeded` reveals that the components referenced in the routes.jsx like `QuestionnaireEditor`and `SurveySimulation` are the ones responsibles to load from params into the redux if needed. Leaving the `QuestionnaireTitle` to grab the quex via `withQuestionnaire`.

Is there any issue to move `fetchQuestionnaireIfNeeded` into the `QuestionnaireSimulation`?

File: web/static/js/components/simulation/SimulationSteps.jsx [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+      return step.type == 'section'
+        ? !!step.steps && step.steps.some(st => submissions.some(su => su.stepId == st.id))
+        : submissions.some(su => su.stepId == step.id)
+    }
+
+    const wasSkipped = (step: Step, isActive: boolean, isCompleted: boolean, headedBySectionSteps: Array<Step>): boolean => {
+      if (isActive || isCompleted) return false
+      if (simulationIsEnded) {
+        // if the simulation is ended (and the step is neither active nor completed)
+        // then, it was skipped
```
Comment (bcardiff) [inline web/static/js/components/simulation/SimulationSteps.jsx:67]: Append comment that section will be marked as completed because isCompleted will return true.

Otherwise it seems that section will be marked as skipped with simulation ended.

File: web/static/js/components/simulation/SimulationSteps.jsx [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+          step.steps
+        ]
+      } else {
+        return step
+      }
+    }, 2)
+    return {
+      headedBySectionSteps: headedBySectionSteps.map(step => ({
+        ...step,
+        status: (
```
Comment (bcardiff) [inline web/static/js/components/simulation/SimulationSteps.jsx:94]: extract to stepStatus function

File: web/static/js/components/questionnaires/QuestionnaireTitle.jsx (hunk 6 above)
Comment (aeatencio) [reply web/static/js/components/questionnaires/QuestionnaireTitle.jsx:74]: Avoid fetching the questionnaire from `QuestionnaireSimulation` was done intentionally [here](https://github.com/instedd/surveda/pull/1698/commits/40b62e16f91fc789f5458a87bcfc0bb2b2559922).
The intention was to avoid fetching something that the component doesn't need for itself.
But I can revert that with no problem if you think it was better that way.

Comment (bcardiff) [reply web/static/js/components/questionnaires/QuestionnaireTitle.jsx:74]: I think is best to keep the same design: allow the top-level component to do the fetch from the url. Otherwise the components like title end up depending on the url, instead of receiving props.

File: web/static/js/components/simulation/SimulationSteps.jsx (hunk 8 above)
Comment (aeatencio) [reply web/static/js/components/simulation/SimulationSteps.jsx]: Done!

File: web/static/js/components/simulation/SimulationSteps.jsx (hunk 7 above)
Comment (aeatencio) [reply web/static/js/components/simulation/SimulationSteps.jsx]: I modified how this is expressed in [this commit](https://github.com/instedd/surveda/pull/1698/commits/75372fcd8aacf8f74973733085179c0ac2b73b98), in the context of the `stepStatus` extraction.

File: web/static/js/components/questionnaires/QuestionnaireTitle.jsx (hunk 6 above)
Comment (aeatencio) [reply web/static/js/components/questionnaires/QuestionnaireTitle.jsx:74]: OK, so I'll revert that change. 👍

Comment (aeatencio) [reply web/static/js/components/questionnaires/QuestionnaireTitle.jsx]: Done!

**Comment by aeatencio** [comment]
Again, @bcardiff, ready for review! 🤞
